ENH: Add experimental visual DICOM browser

Punzo opened the pull request:
**:zap: Updated[^1] Description: :zap:**

[^1]: _The description has been updated by @jcfr on 2024.01.19 based on the consolidated list of commits integrating the CTK changes merged through https://github.com/commontk/CTK/pull/1165_


This enhancement includes an update to CTK and introduces the `ctkDICOMVisualBrowserWidget` to augment DICOM exploration, query, and retrieval capabilities within the DICOM module. The widget is seamlessly integrated into the Slicer DICOM module and is labeled as experimental.

For a comprehensive overview and future plans, refer to the roadmap at:
* https://github.com/commontk/CTK/issues/1162.

By default, the widget is disabled, and users can access the familiar `ctkDICOMBrowser` when opening the `DICOM` module. To enable the experimental feature, users can toggle the option in the dropdown menu of the `Show DICOM database` pushbutton in the `DICOM` module UI.

In `SlicerDICOMBrowser`, this introduces an instance of `ctkDICOMVisualBrowserWidget` alongside the existing `ctkDICOMBrowser` instance. When users activate the experimental feature, widget visibilities are adjusted accordingly. Both widgets share the same DICOM folder set in the DICOM settings.

List of CTK changes:

```
$ git shortlog 45f33c81..88ff72b9 --group=author --group=trailer:co-authored-by --no-merges
Andras Lasso (1):
      ENH: Add Visual DICOM Browser (#1165)

Davide Punzo (1):
      ENH: Add Visual DICOM Browser (#1165)

Jean-Christophe Fillion-Robin (1):
      ENH: Add Visual DICOM Browser (#1165)
```

-----------------

**Original Description:**

@lassoan @pieper @jcfr 
This PR adds the visual DICOM browser widget. CTK reference PR https://github.com/commontk/CTK/pull/1142
For more info see CTK PR [overview](https://github.com/commontk/CTK/pull/1142#issuecomment-1759806374) .

We tested the widget on Linux and Windows on multiple servers (e.g. Orthanc and [www.dicomserver.co.uk](http://www.dicomserver.co.uk/)), but we are flagging this feature as experimental.

The developement is still ongoing (see https://github.com/commontk/CTK/issues/1162). However, I would like to start getting feedback and if anyone can test on MacOS or different server configurations would be highly welcome.

In default I have set the widget hidden and the user has to click the advanced menu for accessing it. See below:
[Screencast from 2024-01-09 14-28-57.webm](https://github.com/Slicer/Slicer/assets/7985338/b1df0862-cc29-42b4-85dc-f9bec2da189a)

lassoan reviewed (approved):
This works quite nicely - we are getting very close! We could merge this PR as is, but there are a number of small regressions/annoyances compared to the old browser which we should fix before asking people to start testing (see below). Let us know if you prefer to merge this as is and submit a new PR with fixes or you want to include fixes in this PR.

Higher priority (needs to be fixed before we ask users to start testing):
- [ ] Hide "Actions" section (Close and Import buttons)
- [ ] There is some custom theming that messes up the display: text is very large, and colors are not correct when the application is in dark mode (text is not visible in search fields at all, because background of the fields are white or yellow and text color is white; it should be all standard background color). One technique that can be used is to detect if text is white or black and adjust default colors accordingly (see `qMRMLSubjectHierarchyTreeViewPrivate::updateColors()`). Things will be simpler if Slicer will use QSS for color themes, but we are not there yet - this can be another topic to work on during the project week (with @sjh26).
- [ ] When I switched to the new browser, I got a message that the database is not compatible. I chose to create a new database, the message went away, I don't get any error messages, but I did not see anything in query results. In the application log I see "[Qt] QSqlQuery::prepare: database not open". It may be possible that I just did not click "Apply changes" in settings after enabling the UK DICOM server. Restarting the application and enabling the UK DICOM server solved it (I don't know if the restart was needed or not, it would be nice if you could verify if transition from an old database schema works well - no application restart should be needed).
- [ ] Attempting to load the "XA" series in "Rubo DEMO" patient (which fails with some error that it cannot retrieve the data), the DICOM browser does not work anymore: changing the query parameters and hitting the search button has no effect, the application is still stuck on the same patient. I cannot even close the application, because the exit confirmation popup gets immediately hidden.
- [ ] Choice of "Show experimental visual DICOM browser" setting should be preserved when the application is restarted. It can be a Boolean flag in DICOM section in the application settings.
- [ ] Thumbnail size would need to be about 1/2 or 1/3. I could be nice if it is configurable in Settings section (at least small/medium/large, but maybe with a slider; if could be maybe specified as percentage of window height)
- [ ] If "Auto-hide browser window" option is enabled, then the view layout should change automatically to show the loaded image.
- [ ] Show local database content by default (when showing the database without specifying anything).
- [ ] Loading error/warning messages must be displayed the same way as for the old browser. For example, clicking on "RS:..." series in "Ghasemi" patient data set has simply no effect.

Lower priority (not needed before integration):
- [ ] While waiting for download of a complete sequence, we can still do some user actions, but not all. I suspect that it is because we pump the event loop with `app->processEvents()`, but that might be unintentional. We might want to call `processEvents` with flags that disable user interaction, or even better, show a model popup with a "Cancel" button.
- [ ] We need to expose the "Advanced" loading option, when a list of loadables is displayed when loading is requested and user can choose what item(s) to load.
- [ ] I don't get some updates (study list remains empty) when setting filter criteria to "Everything from last year". I see "ctkDICOMQuery: the number of responses of the query task at patients level surpassed the maximum value of permitted results (i.e. 25)." message in the log - can it be the root cause? Or maybe just that this filter does not filter out patients. It is confusing that some filter boxes filter patients, while others let all patients displayed, even those that don't have matching studies/series/modality/date.
- [ ] Having a shortcut to browser previous/next patient (Ctrl-tab/Ctrl-shift-tab) would be useful when working with a research PACS (where we are not find/explore data, without necessarily knowing patient name/id).
- [ ] Thumbnails are quite low quality, which is visible when the image contains text or simple graphics - if there is an option to use higher-order interpolation and it is not much slower then it would be nice to switch to that

Punzo commented:
> This works quite nicely - we are getting very close! We could merge this PR as is, but there are a number of small regressions/annoyances compared to the old browser which we should fix before asking people to start testing (see below). Let us know if you prefer to merge this as is and submit a new PR with fixes or you want to include fixes in this PR.
> 
> Higher priority (needs to be fixed before we ask users to start testing):
> 
> * [ ]  Hide "Actions" section (Close and Import buttons)
> * [ ]  There is some custom theming that messes up the display: text is very large, and colors are not correct when the application is in dark mode (text is not visible in search fields at all, because background of the fields are white or yellow and text color is white; it should be all standard background color). One technique that can be used is to detect if text is white or black and adjust default colors accordingly (see `qMRMLSubjectHierarchyTreeViewPrivate::updateColors()`). Things will be simpler if Slicer will use QSS for color themes, but we are not there yet - this can be another topic to work on during the project week (with @sjh26).
> * [ ]  When I switched to the new browser, I got a message that the database is not compatible. I chose to create a new database, the message went away, I don't get any error messages, but I did not see anything in query results. In the application log I see "[Qt] QSqlQuery::prepare: database not open". It may be possible that I just did not click "Apply changes" in settings after enabling the UK DICOM server. Restarting the application and enabling the UK DICOM server solved it (I don't know if the restart was needed or not, it would be nice if you could verify if transition from an old database schema works well - no application restart should be needed).
> * [ ]  Attempting to load the "XA" series in "Rubo DEMO" patient (which fails with some error that it cannot retrieve the data), the DICOM browser does not work anymore: changing the query parameters and hitting the search button has no effect, the application is still stuck on the same patient. I cannot even close the application, because the exit confirmation popup gets immediately hidden.
> * [ ]  Choice of "Show experimental visual DICOM browser" setting should be preserved when the application is restarted. It can be a Boolean flag in DICOM section in the application settings.
> * [ ]  Thumbnail size would need to be about 1/2 or 1/3. I could be nice if it is configurable in Settings section (at least small/medium/large, but maybe with a slider; if could be maybe specified as percentage of window height)
> * [ ]  If "Auto-hide browser window" option is enabled, then the view layout should change automatically to show the loaded image.
> * [ ]  Show local database content by default (when showing the database without specifying anything).
> * [ ]  Loading error/warning messages must be displayed the same way as for the old browser. For example, clicking on "RS:..." series in "Ghasemi" patient data set has simply no effect.
> 
> Lower priority (not needed before integration):
> 
> * [ ]  While waiting for download of a complete sequence, we can still do some user actions, but not all. I suspect that it is because we pump the event loop with `app->processEvents()`, but that might be unintentional. We might want to call `processEvents` with flags that disable user interaction, or even better, show a model popup with a "Cancel" button.
> * [ ]  We need to expose the "Advanced" loading option, when a list of loadables is displayed when loading is requested and user can choose what item(s) to load.
> * [ ]  I don't get some updates (study list remains empty) when setting filter criteria to "Everything from last year". I see "ctkDICOMQuery: the number of responses of the query task at patients level surpassed the maximum value of permitted results (i.e. 25)." message in the log - can it be the root cause? Or maybe just that this filter does not filter out patients. It is confusing that some filter boxes filter patients, while others let all patients displayed, even those that don't have matching studies/series/modality/date.
> * [ ]  Having a shortcut to browser previous/next patient (Ctrl-tab/Ctrl-shift-tab) would be useful when working with a research PACS (where we are not find/explore data, without necessarily knowing patient name/id).
> * [ ]  Thumbnails are quite low quality, which is visible when the image contains text or simple graphics - if there is an option to use higher-order interpolation and it is not much slower then it would be nice to switch to that

thanks for testing and the feedback.

I have added the lower priority to https://github.com/commontk/CTK/issues/1162
I am working now on the higher priority ones.

Punzo commented:
High priority fix.
Reference: https://github.com/commontk/CTK/pull/1163

- [ ] Hide "Actions" section (Close and Import buttons)
- [ ]  There is some custom theming that messes up the display: text is very large, and colors are not correct when the application is in dark mode (text is not visible in search fields at all, because background of the fields are white or yellow and text color is white; it should be all standard background color). One technique that can be used is to detect if text is white or black and adjust default colors accordingly (see qMRMLSubjectHierarchyTreeViewPrivate::updateColors()). Things will be simpler if Slicer will use QSS for color themes, but we are not there yet - this can be another topic to work on during the project week (with @sjh26).
- [ ]  When I switched to the new browser, I got a message that the database is not compatible. I chose to create a new database, the message went away, I don't get any error messages, but I did not see anything in query results. In the application log I see "[Qt] QSqlQuery::prepare: database not open". It may be possible that I just did not click "Apply changes" in settings after enabling the UK DICOM server. Restarting the application and enabling the UK DICOM server solved it (I don't know if the restart was needed or not, it would be nice if you could verify if transition from an old database schema works well - no application restart should be needed).
- [ ]  Attempting to load the "XA" series in "Rubo DEMO" patient (which fails with some error that it cannot retrieve the data), the DICOM browser does not work anymore: changing the query parameters and hitting the search button has no effect, the application is still stuck on the same patient. I cannot even close the application, because the exit confirmation popup gets immediately hidden.
- [ ]  Choice of "Show experimental visual DICOM browser" setting should be preserved when the application is restarted. It can be a Boolean flag in DICOM section in the application settings.
- [ ]  Thumbnail size would need to be about 1/2 or 1/3. I could be nice if it is configurable in Settings section (at least small/medium/large, but maybe with a slider; if could be maybe specified as percentage of window height)
- [ ]  If "Auto-hide browser window" option is enabled, then the view layout should change automatically to show the loaded image.
- [ ]  Show local database content by default (when showing the database without specifying anything).
- [ ]  Loading error/warning messages must be displayed the same way as for the old browser. For example, clicking on "RS:..." series in "Ghasemi" patient data set has simply no effect.

Punzo commented:
@lassoan I have applied all the high priority comments + few other from https://github.com/commontk/CTK/issues/1162

Please have a look again, thanks!

lassoan commented:
Amazing, thank you, I'll test it and if things look good then we get these integrated into CTK and Slicer by tomorrow.

Punzo commented:
> Amazing, thank you, I'll test it and if things look good then we get these integrated into CTK and Slicer by tomorrow.

you are welcome and thanks! btw: I just force pushed a couple of minor fixes to https://github.com/commontk/CTK/pull/1163. I will wait for your testing 🤞.

lassoan commented:
This works very nicely, it should be ready to go, once the CTK update is integrated. Don't forget to update the CTK git hash.

jcfr commented:
* Suggest a more descriptive commit message explaining the changes and impact on the user
* Prepare a discourse post describing the feature with link to _issue roadmap_

lassoan commented:
> * Suggest a more descriptive commit message explaining the changes, impact on the user and how to enable the experimental feature.
> * Prepare a discourse post describing the feature with link to _issue roadmap_

Are these tasks for @Punzo ?

jcfr commented:
> > * Suggest a more descriptive commit message explaining the changes, impact on the user and how to enable the experimental feature.
> > * Prepare a discourse post describing the feature with link to _issue roadmap_
> 
> Are these tasks for @Punzo ?

That would be great.

The discourse post could be drafted and collaboratively worked on here:
https://hackmd.io/@2gubbV2dS1CjOx_adMoxNg/HkIz2K7t6

Examples of posts are available here:
https://discourse.slicer.org/search?expanded=true&q=tags%3Afeature%20order%3Alatest_topic

Punzo commented:
Proposed commit message:

```
Add Visual DICOM Browser

This commit introduces the `ctkDICOMVisualBrowserWidget` to enhance DICOM exploration, query, and retrieval within the DICOM module. The widget is marked as experimental, and a comprehensive roadmap is available at https://github.com/commontk/CTK/issues/1162.

By default, the widget is disabled, and users can access the previous `ctkDICOMBrowser` when opening the `DICOM` module. To enable the experimental feature, users can toggle the option in the dropdown menu of the `Show DICOM database` pushbutton in the `DICOM` module UI.

In `SlicerDICOMBrowser`, this commit adds an instance of `ctkDICOMVisualBrowserWidget` alongside the existing `ctkDICOMBrowser` instance. When users toggle the experimental feature, widget visibilities are adjusted accordingly. Both widgets share the same DICOM folder set in the DICOM settings.

Co-authored-by: Andras Lasso <lasso@queensu.ca>
Co-authored-by: Jean-Christophe Fillion-Robin <jchris.fillionr@kitware.com>
```

Punzo commented:
@lassoan @jcfr I have edited https://hackmd.io/SFrAGE4CS8uSkI2YvDRKRg?both . Please have a look, thanks!

Punzo commented:
@jcfr ops this commit https://github.com/Slicer/Slicer/pull/7525/commits/c223bb898118d88b3c29a284451537297a51e503 
 (based on https://github.com/commontk/CTK/pull/1165/commits/1b84e8ce138b9a06f20962e861e2d06d1d0a5be7) gives error. I am fixing it now. Thanks for the patience.

Punzo commented:
> @jcfr ops this commit [c223bb8](https://github.com/Slicer/Slicer/commit/c223bb898118d88b3c29a284451537297a51e503) (based on [commontk/CTK@1b84e8c](https://github.com/commontk/CTK/commit/1b84e8ce138b9a06f20962e861e2d06d1d0a5be7)) gives error. I am fixing it now. Thanks for the patience.

ok done, fixed in https://github.com/Slicer/Slicer/pull/7525/commits/d7bc9207786e56feaa4d58d68007021ea4e72ff6
